File: Lighthouse/LHBasicCommandRegistry.m
//
//  LHBasicCommandRegistry.m
//  Lighthouse
//
//  Created by Nick Tymchenko on 17/09/15.
//  Copyright (c) 2015 Pixty. All rights reserved.
//

#import "LHBasicCommandRegistry.h"
#import "LHCommand.h"
#import "LHTarget.h"

@interface LHBasicCommandRegistry ()

@property (nonatomic, strong) NSMapTable *blocksByCommandClass;

@end


@implementation LHBasicCommandRegistry

#pragma mark - Init

- (instancetype)init {
    self = [super init];
    if (!self) return nil;
    
    _blocksByCommandClass = [NSMapTable strongToStrongObjectsMapTable];
    
    return self;
}

#pragma mark - Setup

- (void)bindCommandClass:(Class)commandClass toTarget:(LHTarget *)target {
    [self bindCommandClass:commandClass toBlock:^LHTarget *(id<LHCommand> command) {
        return target;
    }];
}

- (void)bindCommandClass:(Class)commandClass toTargetWithActiveNode:(id<LHNode>)node {
    [self bindCommandClass:commandClass toTarget:[LHTarget withActiveNode:node]];
}

- (void)bindCommandClass:(Class)commandClass toTargetWithInactiveNode:(id<LHNode>)node {
    [self bindCommandClass:commandClass toTarget:[LHTarget withInactiveNode:node]];
}

- (void)bindCommandClass:(Class)commandClass toTargetWithActiveNode:(id<LHNode>)node origin:(LHRouteHintOrigin)origin {
    LHRouteHint *hint = [[LHRouteHint alloc] initWithNodes:nil edges:nil origin:origin];
    [self bindCommandClass:commandClass toTarget:[LHTarget withActiveNode:node routeHint:hint]];
}

- (void)bindCommandClass:(Class)commandClass toTargetWithInactiveNode:(id<LHNode>)node origin:(LHRouteHintOrigin)origin {
    LHRouteHint *hint = [[LHRouteHint alloc] initWithNodes:nil edges:nil origin:origin];
    [self bindCommandClass:commandClass toTarget:[LHTarget withInactiveNode:node routeHint:hint]];
}

- (void)bindCommandClass:(Class)commandClass toBlock:(LHCommandTargetProvidingBlock)block {
    [self.blocksByCommandClass setObject:[block copy] forKey:commandClass];
}

#pragma mark - LHCommandRegistry

- (LHTarget *)targetForCommand:(id<LHCommand>)command {
    LHCommandTargetProvidingBlock block = [self.blocksByCommandClass objectForKey:[command class]];
    return block ? block(command) : nil;
}

@end

File: Lighthouse/LHBasicCommandRegistry.swift
//
//  LHBasicCommandRegistry.swift
//  Lighthouse
//
//  Created by Artem Starosvetskiy on 06.02.2021.
//

public extension LHBasicCommandRegistry {
    func bind<Command>(_ commandType: Command.Type, to target: LHTarget) where Command: LHCommand {
        __bindCommand(commandType, to: target)
    }

    func bind<Command>(_ commandType: Command.Type, activating node: LHNode) where Command: LHCommand {
        __bindCommand(commandType, to: LHTarget.withActiveNode(node))
    }

    func bind<Command>(_ commandType: Command.Type, deactivating node: LHNode) where Command: LHCommand {
        __bindCommand(commandType, to: LHTarget.withInactiveNode(node))
    }

    func bind<Command>(_ commandType: Command.Type, activating node: LHNode, origin: LHRouteHintOrigin) where Command: LHCommand {
        let hint = LHRouteHint(nodes: nil, origin: origin, bidirectional: false)
        return __bindCommand(commandType, to: LHTarget.withActiveNode(node, routeHint: hint))
    }

    func bind<Command>(_ commandType: Command.Type, deactivating node: LHNode, origin: LHRouteHintOrigin) where Command: LHCommand {
        let hint = LHRouteHint(nodes: nil, origin: origin, bidirectional: false)
        return __bindCommand(commandType, to: LHTarget.withInactiveNode(node, routeHint: hint))
    }

    func bind<Command>(_ commandType: Command.Type, to targetFactory: @escaping (Command) -> LHTarget) where Command: LHCommand {
        __bindCommand(commandType, to: { command in
            guard let command = command as? Command else { return nil }

            return targetFactory(command)
        })
    }
}
